### PR TITLE
Fix serialization bug in allocation explain API.

### PR DESCRIPTION
**Pull request by martijnvg**
The allocation explain API would fail if the request was executed via a non elected master node. This triggers the request to be redirected to the elected master node. The internal response serialization would fail because the read code would use regular ints, while the write code uses  vints.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/ClusterInfo.java [hunk 1]
```diff
@@ -68,25 +68,25 @@ public ClusterInfo(ImmutableOpenMap<String, DiskUsage> leastAvailableSpaceUsage,
     }
 
     public ClusterInfo(StreamInput in) throws IOException {
-        int size = in.readInt();
+        int size = in.readVInt();
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/ClusterInfo.java]: I wonder if we should introduce a method to InputStream/OutputStream to read maps. We have similar methods for lists. Something like:
```
<K, V> Map<K, V> readMap(Reader<K> keyReader, Reader<V> valueReader)
```

I can think of about a dozen places we could use it.

File: core/src/test/java/org/elasticsearch/cluster/ClusterInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        ClusterInfo result = new ClusterInfo(output.bytes().streamInput());
+        assertEquals(clusterInfo.getNodeLeastAvailableDiskUsages(), result.getNodeLeastAvailableDiskUsages());
+        assertEquals(clusterInfo.getNodeMostAvailableDiskUsages(), result.getNodeMostAvailableDiskUsages());
+        assertEquals(clusterInfo.shardSizes, result.shardSizes);
+        assertEquals(clusterInfo.routingToDataPath, result.routingToDataPath);
+    }
+
+    private static ImmutableOpenMap<String, DiskUsage> randomDiskUsage() {
+        int numEntries = randomIntBetween(0, 128);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/cluster/ClusterInfoTest.java]: I wonder if you should do something like `+ rarely() ? 1000 : 0`. Just to be sure we hit the case where vint is different.

Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/cluster/ClusterInfoTest.java]: `randomIntBetween(0, 128)` does return `0` at some point in time, is that sufficient?

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/cluster/ClusterInfoTest.java]: Actually it should be fine now that I think about it. Ignore that comment!

File: core/src/main/java/org/elasticsearch/cluster/ClusterInfo.java (hunk 1 above)
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/cluster/ClusterInfo.java]: Makes sense. I've added that.

**Comment by dakrone** [comment]
Thanks for fixing this @martijnvg!

**Comment by jpountz** [comment]
@martijnvg can you update the description of this PR to explain what was broken?

**Comment by martijnvg** [comment]
@jpountz Done, does this look better?

**Comment by jpountz** [comment]
Thanks @martijnvg !

**Pull request merged**